pkg/controller/mizar/node_controller.go

```diff
@@ -0,0 +1,307 @@
+		klog.Fatalln(c.logInfoMessage("Timeout expired during waiting for caches to sync."))
+	}
+	klog.Infoln(c.logInfoMessage("Starting workers..."))
+	for i := 0; i < workers; i++ {
+		go wait.Until(c.runWorker, time.Second, stopCh)
+	}
+	<-stopCh
+	klog.Info(c.logInfoMessage("Shutting down node controller"))
+}
+
+// Enqueue puts key of the node object in the work queue
+func (c *NodeController) Enqueue(command string, key string) {
+	commandKey := c.genCommandKey(command, key)
+	c.queue.Add(commandKey)
+}
+
+// Dequeue an item and process it
+func (c *NodeController) runWorker() {
+	for {
+		item, queueIsEmpty := c.queue.Get()
+		if queueIsEmpty {
+			break
+		}
+		c.process(item)
+	}
+}
+
+// Parsing a item key and call gRPC request
+func (c *NodeController) process(item interface{}) {
+	defer c.queue.Done(item)
+	commandKey, ok := item.(string)
+	if !ok {
+		klog.Errorf(c.logInfoMessage("unexpected item in queue: " + commandKey))
+		c.queue.Forget(item)
+		return
+	}
+	command, key, err := c.parseCommandKey(commandKey)
+	if err != nil {
+		klog.Errorf(c.logInfoMessage("Unexpected item in queue: " + commandKey))
+		c.queue.Forget(item)
+		return
+	}
+	nodeKind, nodeTenant, nodeName, nodeStatusType, nodeStatus, nodeAddressType, nodeAddress, err := c.parseKey(key)
+	if err != nil {
+		klog.Errorf(c.logInfoMessage("Unexpected string in queue; discarding: ") + key)
+		c.queue.Forget(item)
+		return
+	}
+	if nodeKind != NODE_KIND {
+		klog.Errorf(c.logInfoMessage("Unexpected object in queue; discarding: ") + key)
+		c.queue.Forget(item)
+		return
+	}
+	klog.V(5).Infof(c.logInfoMessage("Processing a node: ")+"%s/%s/%s/%s/%s/%s/%s", nodeKind, nodeTenant, nodeName, nodeStatusType, nodeStatus, nodeAddressType, nodeAddress)
+	c.gRPCRequest(command, nodeAddress)
+	c.queue.Forget(item)
+}
+
+// Generate a node key
+func (c *NodeController) genKey(resource *v1.Node) string {
+	nodeKind := resource.GetObjectKind().GroupVersionKind().Kind
+	if nodeKind == "" {
+		nodeKind = NODE_KIND
+	}
+	nodeTenant := resource.GetTenant()
+	nodeName := resource.GetName()
+	if nodeName == "" {
+		klog.Fatalln(c.logInfoMessage("Node name should not be null"))
+	}
+	var nodeStatusType, nodeStatus string
+	conditions := resource.Status.Conditions
+	if len(conditions) < 6 {
+		klog.Fatalln(c.logInfoMessage("Node status information is needed"))
+	} else {
+		nodeStatusType = fmt.Sprintf("%s", resource.Status.Conditions[5].Type)
+		nodeStatus = fmt.Sprintf("%s", resource.Status.Conditions[5].Status)
+	}
+	var nodeAddressType, nodeAddress string
+	addresses := resource.Status.Addresses
+	if len(addresses) < 1 {
+		klog.Fatalln(c.logInfoMessage("Node ip address is needed"))
+	} else {
+		nodeAddressType = fmt.Sprintf("%s", resource.Status.Addresses[0].Type)
+		nodeAddress = fmt.Sprintf("%s", resource.Status.Addresses[0].Address)
+	}
+	key := fmt.Sprintf("%s/%s/%s/%s/%s/%s/%s", nodeKind, nodeTenant, nodeName, nodeStatusType, nodeStatus, nodeAddressType, nodeAddress)
```

Comment:
it is very rare the key used in queue to have very detailed info. The key is for retrieving item out of the local storage; usually a few values(typiclly node name in this case, seemingly) would suffice.

Reply:
Thank you for the helpful comment. I would like to ask your opinion. 
Firstly, I agree with you and I do as you suggested for other controllers. 
But, this is a little bit exceptional because only information this controller needs is "IP address" . 
For the IP only, it may be more efficient to keep IP with Key. Otherwise it needs to take the object from storage and reprocess object data to get IP address. Could you please share your opinion?